### PR TITLE
Enable generic tuples; improve enum const inits

### LINKED ISSUE #7681: Scoped enum offset issues
### Summary of Problem

A few uses of a scoped enum with an offset fail to compile, when I
think they should work.

### Steps to Reproduce

**Source Code:**

[ edit 7/19/2024 cassella -- only the last of these examples still fails. ]

```chapel
enumScopeTest(2);
enumScopeTest(3);
proc enumScopeTest(param x) {
  enum Numbers {one=x, two, three};
  writeln(Numbers.one);
}
```

```
scope-param-offset.chpl:5: error: enumerator 'one' is not an integer param value
```

Maybe it needs to be reassured that ``x`` is an int param?

```chapel
enumScopeTest(2);
enumScopeTest(3);
proc enumScopeTest(param x:int) {
  enum Numbers {one=x, two, three};
  writeln(Numbers.one);
}
```

```
scope-param-offset-type-2.chpl:5: error: enumerator 'one' is not an integer param value
```

Try to impart this reassurance a different way?

```chapel
enumScopeTest(2);
enumScopeTest(3);
proc enumScopeTest(param x) {
  enum Numbers {one=x:int, two, three};
  writeln(Numbers.one);
}
```

```
scope-param-offset-type.chpl:5: internal error: CAL0120 chpl Version 1.17.0 pre-release (cf54f6a)
...
```

It also doesn't work if the ``param x`` is global (and the function
the enum is within is now concrete):

```chapel
enumScopeTest();
param x = 2;
proc enumScopeTest() {
  enum Numbers {one=x, two, three};
  writeln(Numbers.one);
}
```

```
$CHPL_HOME/modules/standard/Types.chpl:607: internal error: TYP0356 chpl Version 1.17.0 pre-release (cf54f6a)
```

It works if the ``enum`` and ``writeln`` are at file scope with ``x``.

Edit 3/15/2019: The error for this is now:
```
internal error: Unable to resolve enumerator type expression [resolution/functionResolution.cpp:6183]
```

Edit, part 2: I also noticed that if the call to `enumScopeTest()` is
moved to after the function's definition, or even to just after the
`param x = 2` line, it compiles and runs.

FWIW, the spec doesn't spell out in ``7.2 Enumerated Types`` that the
``init-part`` expression has to be a param.

**Associated Future Test(s):**

[`test/types/enum/scope-param-offset-type.chpl`](
https://github.com/chapel-lang/chapel/blob/master/test/types/enum/scope-param-offset-type.chpl
)
[`test/types/enum/scope-param-offset-type-2.chpl`](
https://github.com/chapel-lang/chapel/blob/master/test/types/enum/scope-param-offset-type-2.chpl
)
[`test/types/enum/scope-param-offset-concrete.chpl`](
https://github.com/chapel-lang/chapel/blob/master/test/types/enum/scope-param-offset-concrete.chpl
)
[`test/types/enum/scope-param-offset.chpl`](
https://github.com/chapel-lang/chapel/blob/master/test/types/enum/scope-param-offset.chpl
)

[edit 11/1 cassella: add links to futures ]

### Configuration Information

This is with ``chpl`` built from master late Friday,

```
fortytwo@magrathea:~/src/chapel/test/types/enum (generic-function-enums)$ chpl --version 
chpl Version 1.17.0 pre-release (cf54f6a)
Copyright (c) 2004-2017, Cray Inc.  (See LICENSE file for more details)
fortytwo@magrathea:~/src/chapel/test/types/enum (generic-function-enums)$ printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_MAKE: make
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_AUX_FILESYS: none
fortytwo@magrathea:~/src/chapel/test/types/enum (generic-function-enums)$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
```

### LINKED ISSUE #11461: compiler rejects generic declared yield type
While #10184 is being addressed, the same issue with iterators still remains.

Tests:
```
test/users/thom/yieldGeneric.chpl
test/functions/vass/declaredGenericReturnTuple.chpl
```

Cf. this works: `test/functions/lydia/declaredGenericReturnType.chpl`

### PR DESCRIPTION
**Pull request by vasslitvinov**
Resolves #11461
Improves #7681, #14148

Prior to this PR, tuples could not be generic,
and some initializations of enum constants were not handled well.

With this PR, the compiler handles the following cases correctly.

### A tuple containing generic type(s) is now a generic type

* extends `test/functions/vass/declaredGenericReturnTuple.chpl`

### An iterator declares a generic return/yield type

* Resolves #11461
  - un-futurizes `test/users/thom/yieldGeneric.chpl`
  - un-futurizes `test/functions/vass/declaredGenericReturnTuple.chpl`
  - extends `test/functions/lydia/declaredGenericReturnType.chpl`
    mentioned in #10172

* Makes progress in #14148
  - un-futurizes `test/classes/nilability/errors/cast-nil-to-generic-6.chpl`
  - next steps: add more testing as suggested in
    https://github.com/chapel-lang/chapel/issues/14148#issuecomment-1113755444

### A recursive iterator declares a generic return/yield type

* Eliminates a compiler crash when the iterator is recursive
  - adds `test/functions/iterators/recursive/recursive-generic-declared-type.chpl`
  - next steps: have the error message indicate the problem directly similar to `test/studies/adventOfCode/2021/bradc/futures/day18-bug.chpl`

### A recursive proc declares a generic return type

* Eliminates a compiler crash in this case
  - adds `test/functions/compilation-errors/recursive-generic-return.chpl`
  - next steps: have the error message indicate recursive-ness?

### A constant of a function-scoped enum is initialized with a param

* The param may be the function's formal or a global

* Resolves two cases in #7681
  - un-futurizes `test/types/enum/scope-param-offset.chpl`
  - un-futurizes `test/types/enum/scope-param-offset-type.chpl`
  - a reasonable error in `test/types/enum/scope-param-offset-concrete.chpl`
  - next steps: see if the latter is an acceptable behavior; if so, un-futurize that test and close the issue #7681

Testing: standard paratest; gasnet paratest over multilocale tests; both include -futures.

### CODE REVIEW

**Comment by vasslitvinov** [comment]
@bradcray fyi

File: compiler/resolution/resolveFunction.cpp [hunk 1]
```diff
@@ -487,11 +487,14 @@ void resolveSpecifiedReturnType(FnSymbol* fn) {
       fn->retType = retType;
     }
 
-    if (fn->isIterator() == true && fn->iteratorInfo == NULL) {
+    if (fn->isIterator() == true && fn->iteratorInfo == NULL &&
```
Comment (daviditen) [inline compiler/resolution/resolveFunction.cpp:490]: As long as you're touching this line anyway could you please remove the unnecessary  `== true`?

File: test/studies/adventOfCode/2021/bradc/futures/day18-retNil4.bad [hunk 2]
```diff
@@ -1,8 +1 @@
-day18-retNil4.chpl:12: internal error: RES-RES-ION-2818 chpl version 1.26.0 pre-release (c71dc4aa3f)
-Note: This source location is a guess.
-
-Internal errors indicate a bug in the Chapel compiler ("It's us, not you"),
-and we're sorry for the hassle.  We would appreciate your reporting this bug --
-please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
-the filename + line number above may be useful in working around the issue.
-
+day18-retNil4.chpl:5: error: generic return type 'C?'
```
Comment (daviditen) [inline test/studies/adventOfCode/2021/bradc/futures/day18-retNil4.bad:1]: If you make my suggested change to the error message above this line will need to change too.

File: test/functions/compilation-errors/recursive-generic-return.good [hunk 3]
```diff
@@ -0,0 +1,2 @@
+recursive-generic-return.chpl:12: warning: RR(int(64))
+recursive-generic-return.chpl:15: error: generic return type 'RR'
```
Comment (daviditen) [inline test/functions/compilation-errors/recursive-generic-return.good:2]: If you make my suggested change to the error message above this line will need to change too.

**Review by daviditen — APPROVED** [review]
Looks great overall! I made a few small suggestions, but otherwise LGTM.

**Pull request merged**